noobaa_sa/s3_client.py

```diff
@@ -102,15 +130,47 @@ def delete_bucket(self, bucket_name, empty_before_deletion=False):
                     f"Bucket {bucket_name} is not empty and will not be deleted. "
                     f"Set empty_before_deletion to True to delete it anyway."
                 )
-                raise e
+                raise BucketNotEmpty(e)
             elif e.response["Error"]["Code"] == "NoSuchBucket":
                 log.warn(f"Bucket {bucket_name} does not exist and cannot be deleted")
+                raise NoSuchBucket(e)
             else:
                 log.error(f"Failed to delete bucket {bucket_name}: {e}")
                 raise e
         else:
             log.info(f"Bucket {bucket_name} deleted successfully")
 
+    def head_bucket(self, bucket_name):
+        """
+        Check if a bucket exists in an S3 account using boto3
+
+        Args:
+            bucket_name (str): The name of the bucket to check
+
+            Returns:
+                bool: True if the bucket exists, False otherwise
+
+        Raises:
+            UnexpectedBehaviour: If an unexpected response is received
+
+        """
+        log.info("Checking if bucket exists via an head_bucket call")
+        try:
+            response = self._boto3_client.head_bucket(Bucket=bucket_name)
+            response_code = int(response["ResponseMetadata"]["HTTPStatusCode"])
+        except ClientError as e:
+            response_code = int(e.response["Error"]["Code"])
+        if response_code == 200:
+            log.info(f"head_bucket call to {bucket_name} returned 200 (OK)")
+            return True
+        elif response_code == 404:
+            log.warn(f"head_bucket call to {bucket_name} returned 404 (Not Found)")
+            return False
+        else:
+            raise UnexpectedBehaviour(
+                f"Unexpected response from head_bucket call to {bucket_name}: {response}"
+            )
+
```

Comment:
To standardize s3 operations in this file, Please return the response directory and user will validate the return codes based on their test need

Reply:
Done in https://github.com/red-hat-storage/noobaa-sa-ci/commit/3c9c0dce3ae21b8c2d2006d988a66e1c8d539af9

Note that boto3 returns a ClientError with a different response dict in the error, with a different path to the response code. In my implementation I'm returning either this or the "successful" dict, and to allow uniform access of the response code, I also add it under the "Code" property in the root of the dict.